### DISCUSSION TITLE
Insane(ly) crazy automation idea?

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#90] [General]
CLIPS 6, embedded.  Potentially, it could eliminate current platform / load-level dependent hard timing.  (If x has happened (atomic or list) and now y comes in, we need to do z (but not if I, j, k, etc)).

https://sourceforge.net/projects/clipsrules/files/CLIPS/6.30/

Worked well enough for shuttle refurbishment, rules / checklists.  IMHO, it would be less work than prolog /smalltalk / lisp, etc, and was always meant to be embedded in other programs, and the kicker is its simplicity for new "experts" (GIGO totally applying, as always in anything).

This would also cut the node.js umbilical.

NASA wrote it, we can all use it.  It was effectively made public domain from the beginnings.  It is also very stable, and more or less static for a while language-wise, and always meant to embed in c programs to provide decision support involving specialized rules.

Platform to platform, load-level based on platform and installation, the data being async, effectively, with unpredictable event times.  CLIPS can handle this WITHOUT THE CLOCK, I would think. Out of order arrival can be easily accounted for as well.

Tie it to a console, and end-user written automation can be done as well to handle routine ops...oh wait, it will be... hmmmmmmmmmmmmmmm..  two ports, one for scripts, one for the console op, add a page for DtCyber-only on the console for end-user rules (with a prominent RTFM notice / liability disclaimer).

Of course all of the jobs done in node would have to be redone...  Isn't gnu wget also a library function?  If not (wondering why), it can be.  Other protocols could be added, even ftam if needed...

It would need EXPLAIN and Plato at some point too, stimulating new stuff there.  A new Plato pack, no matter how sparse, and a redist of layout would get rid of the log,14 "out of the box".

IMHO, the standalone version of CLIPS is for a different paradigm, but free and comes with a good newb manual, and would augment the dist (or provide a link for interested parties, with at least the newb and lrm manuals included in the dist).

It's a Cyber! Let's get weird! ( even old 3000'ers would appreciate ).

In honor of some cool dudes formerly at Point Mugu, by the back door, in a room called "shuttle realtime" (NASA net call "cyber")...  (Holy crap. It flies like bricks don't...Total E-Ticket ride), beautiful, large conrac monitors...  CLIPS ain't directly connected, but, anyone refurbishing that, after that, deserves to be honored as well. They never had a fup.  Hell, in honor of John Young, and Roger Crippen, who showed me the "Genesis Rock", a couple of weeks before the test flight (he found a lot more on 16), hell, to all of those crazy sobs who trusted CLIPS enough to strap themselves on top of one, and trust it to keep them safe up, safe in a molecular vacuum with extreme thermal differentials, and safe down to get them home to tuck the kids and spouse/other in.  That takes pure unadulterated balls.

Just remember rule #1 with such paradigms: GIGO & "the princess and the pea". Don't get it, Bullwinkle covered both (recurring themes).  Most people who have dealt with enterprise level systems, should be aware of both, people who want to be should become familiar so as to avoid.

**Comment by KC5VDJ** [comment]
This has been amended over a few hours.  Reload, please.

**Comment by KC5VDJ** [comment]
The more I look at this, the more I like it.  The configuration management possibilities....especially given incomplete / mismatched documentation, possibly incomplete and / or mismatched products,this could tie it together? which pp capsules are needed for __________?  What has do be done in what order with what results, assuming it's checklist is complete, while building and altering checklists itself?  Especially with the dwindling first-hand knowledge base.  The 3rd conference proceedings clearly show CLIPS is a match for the task.  Consider the MILSTAR terminal network configuration example, among others...

**Reply by wschaub** [reply]
How does this solve timing issues processing keystrokes on the console by the DSD PP program itself? it throws away some of the keystrokes in a very time dependent manner (I think to simulate a sort of debounce)

**Reply by KC5VDJ** [reply]
It is all event based... As long as say a checklist is cool, "did this message happen at some point? Then we can proceed".

There are similarities to prolog's assertions.  Something is asserted, the rules then are checked for a match.  This is just a bit higher level.  It's a very capable 80's expert system, that is still used and current, and embeddable, and public domain (might be bsd blurb, I'll have to check), I will give credit to NASA and current maintainers.

"About CLIPS

Developed at NASA's Johnson Space Center from 1985 to 1996, the 'C' Language Integrated Production System (CLIPS) is a rule-based programming language for creating expert systems and other programs where a heuristic solution is easier to implement than an algorithmic solution.  Written in C for portability, CLIPS can be installed and used on a wide variety of platforms.  Since 1996, CLIPS has been available as public domain software."

There are examples of what has been done in the documentation zip for clips, see the 3rd conference proceedings.  I am convinced on suitability for the task, and I've been looking for such a task that I'm willing to spend the time to do it for for years, and it gives me an excuse to take more than a basic dive into CLIPS.

Potentially it can also be used for even building nos 2.8.7.1, it will happen, somebody will, might as well make it easier now.  The 3rd conference doc makes clear the configuration management capabilities.  That could be built up from the install rules maybe.

It would be a lot easier to deal with than embeddable scheme.  With that the inference engine itself would have to be written from scratch, same for prolog, or smalltalk (despite its power, hello sprint).  CLIPS is the best solution I can think of that checks all the boxes, adding such capability, while keeping as close to the KISS principle as possible (in the end, at least. I am not unaware of the scale of change here, this is why I wish to discuss).  Regardless of toolset, this problem begs for an event tripped rule-based inference engine.  This isn't chatapi, although it can do self-modification of rulesets, that is rigidly defined (think lisp / scheme executing data). This is just an extension of the expert system paradigm jump-started by Feigenbaum's xenophobic paranoia (look at the date), probably more used than Dec's OPSnn(?).  Oh, and I did mention the licensing, stability, and maturity?

Hmmm.  The MILSTAR terminal config application had / has about 100 rules, and about 1000 facts... Given the permutations...  Not bad at all.  I wonder what the justification for the language exemption was in Raytheon's proposal?  It's end-user would have dictated Ada instead of c, as a matter of law, but the law did allow for compelling reasons to justify a waiver.  They did have a remote disconnect issue, the inability to, resulting in very expensive phone calls nobody wanted to claim.  Seals use(ed) them.

Still, as with any such system, rulesets require review.  GIGO followed by princess & the pea is in play.  How do you think a fresh ai becomes neo-nazi in hours?  GIGO, early in the "learning" produces a pronounced and amplified effect later, having colored all later "learning".  I could relate this to certain "social" institutions tied to day of week, and early indoctrination issues, but I won't, but the exact same principles apply.  Like I said, the entire Bullwinkle series revolved around these two concepts, it's what made it funny (someone misunderstands or is fed garbage, followed by snowballing fubar).  Hell, The Tick too, for sure, on that matter.

**Reply by KC5VDJ** [reply]
Here is a good description from the case study of CLIPS's role in JSC Mission Control...

"CLIPS is a data-driven language because it uses data in the form of facts to activate and fire if-then rules, which result in a change of execution state for the computer".

Reads like an Oxford definition of an Expert System / Inference Engine.

**Reply by KC5VDJ** [reply]
Problem:

1). We know what order the install is to be done in.
2). We know what messages indicate completion of task.
3). These messages are async by nature, and timing can not be predicted.
4). These messages may arrive in undefined order among other messages, due to multi-tasking and load.
5). Installs have pre-requisites, and may be prerequisite for other installs, as with install steps.

etc.

I am convinced on suitability.  Further expanded might be able to roll up custom builds at finer grained level than currently supported.  It can hold the big picture for configuration management.

Initially, this will simply replace current functionality, fun stuff later, and will be written with total configuration management being the goal, along with ad-hoc operator rulesets for automating mundane or complex console operations (a mere extension of what it will already be doing, just with ad-hoc rules, no sense in leaving an ad-hoc facility out.  Using it, or even enabling it (ad-hoc) will be optional, but, I would anticipate its use, it's too useful).  I would though only distribute one or two examples for end-user ad-hoc.  Just as a child needs to understand the math before a calculator can be excused in a classroom, people need to RTFM (CDC) and get their feet wet first. Remember GIGO and the princess?

Lastly, anyone doing DtCyber either has to keep legacy running, or is a hobbyist, many of which have some actual level of experience with CDC (chose DtCyber instead of Hercules or simh).  All willing, and most probably able, to learn / relearn historic and / or esoteric languages.  Audience compatibility.  Add the level of public exposure to CDI ads in the day, before the scandal, equating CDC with learning about computers.  Again, an audience with a high willing to learn potential.

CLIPS is also temporally appropriate as well, they come from similar timeframes, keeping things "retro", and there are actually indirect (direct?) linkages to cybers historically.  Given that they only made it public-domain in '96 after eleven years of development and use within government and known cyber customers within government and government contractors, a direct linkage to cybers might actually be somewhere.  The NASA endorsement (origin) was / is powerful.  Just ask ETA about the power of a NASA endorsement.  The uses it is / was put to, as per the 3rd conference at JSC, is even more convincing.

**Reply by KC5VDJ** [reply]
As for the "debounce", well, they bought the cheapest ram chips that they could lay hands on, nand gates might have been too expensive.  Also, it would actually be kinda compat with the interface the way node.js is doing things, just doing things a bit different.

**Comment by KC5VDJ** [comment]
[third_clips_conference_proceedings_041059.pdf](https://github.com/kej715/DtCyber/files/12570852/third_clips_conference_proceedings_041059.pdf)

**Comment by KC5VDJ** [comment]
I've made the case.  Who is willing?  This will be my baby, it is my excuse to deep dive CLIPS.

I'm not a cowboy, this requires input, throughout, from the entire DtCyber community, especially contributors.

Fortunately it is a simple language to learn, even beginners can be productive in it quickly, should they so choose.

This will not be a small undertaking, it is a semi-structural change on the host side, again only adding ad-hoc on the cyber operator's side, and will have to be in sync with proposed console changes by its very nature.

I'm sold on it enough to do it.  I hope that I, even as annoying as I am, have convinced all of you.

Please at least browse the linked conference proceedings, and extrapolate.

"Two anchovies, a strawberry creme pie, and a guanaco on water skis."

![cfdb1e4ff78d90cb43538d64f6204b6b--coffee-talk-coffee-coffee](https://github.com/kej715/DtCyber/assets/136194499/db9a26c4-30bb-4195-8138-34775af1b6bf)

**Reply by KC5VDJ** [reply]
I swear Dana Carvey ripped Zippy for the Coffee Klatch sketches.  The non-sequitor was too similar.  Add a "Yow!" to the bit above, it becomes a zippyism.

**Comment by KC5VDJ** [comment]
Followup, it's pure public domain, not even a bsd blurb clause..

"CLIPS License Information

Permission is hereby granted, free of charge, to any person obtaining a 
copy of this software (the "Software"), to deal in the Software without 
restriction, including without limitation the rights to use, copy, modify, 
merge, publish, distribute, and/or sell copies of the Software, and to 
permit persons to whom the Software is furnished to do so.                             

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT OF THIRD PARTY RIGHTS. 
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL 
INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING FROM 
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE 
OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR 
PERFORMANCE OF THIS SOFTWARE."

**Comment by KC5VDJ** [comment]
Hmmm... Picturing a venn set-based representation as the best way to do the configuration management, and to generate the clips data, unless everyone likes tedium.

**Comment by KC5VDJ** [comment]
Might as well, given the LOE.  It'll be worth the LOE though, and simplify so much.

![WINWORD_NC1RmhEoCQ_021605.png](https://github.com/kej715/DtCyber/assets/136194499/169017e7-b437-4bdf-b2dd-415360247125)

**Reply by KC5VDJ** [reply]
(placeholder, correcting)

A first attempt at a first draft.

**Reply by kej715** [reply]
A false premise in the draft is that the current installer is coarse-grained. It is quite finely grained. If you enter the command:

     node install-product list

the installer will display the names of all categories and individual products. You may install whole categories or individual products by name. For example, if you would like to create a system with Algol 68 only, you would enter these commands:

     sudo node install basic
     sudo node install-product algol68

The first command will install a basic system with core products only. The second command adds Algol 68.

If a specifically named product has dependencies, they will be installed too, if they're not already installed.

Also, if a basic of full install fails at any point, it is -not- necessary to start again from scratch. The "continue" keyword can be specified to indicate that you want to continue after the last successfully completed step. For example,

     sudo node install basic continue

or

     sudo node install full continue

**Reply by KC5VDJ** [reply]
Okay, well, first draft. I posted for comments. Will correct.  Thanks.

My problems might have been connectivity related, continue was tried.

**Reply by KC5VDJ** [reply]
What do you think about the paradigm? I see a fit.

**Reply by kej715** [reply]
The proposal to replace the Node.js-based automation with a CLIPS-based one seems to be motivated by the false premise that the timing sensitivity in the automation is due to the Node.js technology itself. The timing sensitivity is due to the original design of the Cyber console and DtCyber's current emulation of it. The Cyber console (e.g., CC545) was a pair of vector displays with no (or very short) phosphor persistence. Consequently, in order to produce a readable display with no flicker, the operating system software was responsible for fully redrawing the displays at least 20 times per second.

PP-based console display drivers such as DSD and DIS have main loops that literally redraw each of the left and right displays as quickly as they can. They issue commands to set X and Y coordinates for each field (not simply each line) to be displayed, and then they send characters for each field. Screen effects such as flashing and high intensity are accomplished by redrawing those fields at different frequencies than the rest of the information on the displays.

DSD and DIS do not assemble screen images in memory before displaying them. They tend to read information from central memory and display it dynamically, field by field. Their main loops re-iterate this dynamic field-by-field drawing behavior just as fast as they can.

DIS is a critical resource in performing software installation on Cyber operating systems. Consider how it operates. DIS reads commands from the keyboard and arranges for the keystrokes entered to be displayed in a field somewhere on one of the displays. It executes channel functions to perform XY positioning and redraw these keystrokes 20+ times per second. While DIS is accepting keystrokes and making them visible on a display, it usually is also displaying the job's dayfile buffer on a display. This is accomplished by reading the dayfile buffer from central memory and executing channel functions to perform XY positioning and character drawing in fields of a display. And, it's re-reading that information and redrawing it, field-by-field, 20+ times per second. This produces a visual that reflects commands that have been entered and the dayfile messages produced by them.

Note that this behavior is quite different from a typical "send/expect" paradigm. It's relatively straightforward to capture keystrokes that have been entered on the console keyboard (i.e., the "send" part), but it's not currently straightforward (or even possible with DtCyber's current emulation of the console) to obtain the direct result of commands entered (i.e., the "expect" part). DIS does not behave like an ordinary user terminal interface.

With an ordinary terminal interface, a user enters a command (the "send" part), and then the system responds by "printing" results back to the terminal (the "expect" part). This isn't how DIS operates. Specifically, the results of entering a command are not "printed" back to the console. The results appear in a dayfile buffer in central memory, and if DIS has been requested to show its *A* display, those results will be visible. Otherwise, they will not. To further complicate the problem, the location in central memory of the dayfile buffer is not predictable because it depends upon the control point to which DIS has been assigned, and that is selected dynamically at the point in time when the DIS command is entered from DSD. The first available control point is usually chosen, and the first available one depends upon how busy the system is and what its configuration currently is.

Even if you know that DIS is currently drawing its *A* display on a particular screen, the XY location at which it will draw a particular dayfile buffer entry is not a constant and will certainly vary from one sequence of commands to another. And it will change as additional commands are entered, or a running program issues additional dayfile message. The current console.c module of DtCyber emulates the console faithfully, meaning that it is responding directly to the channel functions sent by DSD and DIS, and it is fully redrawing the displays just as was done with the original hardware. Even if it exposed the XY positioning and character drawing functions in some way, a consumer would need to handle the enormous volume of information streaming (i.e., due to 20+ cycles/sec refresh), and it would probably need some way of knowing reliably when each refresh cycle was started.

The current Node.js-based automation implementation addresses the problem by using a very simple timing-based approach. It sends a command and simply assumes that the command will complete with expected results within a maximum amount of time.  It's not obvious how CLIPS could solve this problem any differently given the current implementation of DtCyber's console.c module.  Or, if a way is devised for CLIPS to solve it reliably in a different way, why couldn't that solution be applied just as easily to the Node.js-based approach?

The root of the problem is that the original Cyber console design, and DtCyber's current implementation of it, don't behave like an ordinary user terminal interface. In particular, they don't enable the "expect" part of a typical "send/expect" model to be implemented reliably. Solving this problem reliably will almost certainly require enhancing the console.c module, or a new derivative of it, such that "expect" can be specified easily and obtained both predictably and reliably.
